File: blockrunnooor/prompt_bank.py
from __future__ import annotations

import json
import random
from dataclasses import dataclass
from pathlib import Path
from typing import Any


@dataclass(frozen=True)
class PromptItem:
    prompt_id: str
    model: str
    messages: list[dict[str, Any]]
    temperature: float | None
    max_tokens: int | None


def load_prompt_bank(path: str) -> list[PromptItem]:
    p = Path(path)
    if not p.exists():
        return []
    items: list[PromptItem] = []
    for line in p.read_text(encoding="utf-8").splitlines():
        s = line.strip()
        if not s:
            continue
        obj = json.loads(s)
        prompt_id = str(obj.get("prompt_id", "")).strip()
        model = str(obj.get("model", "")).strip()
        messages = obj.get("messages")
        if not prompt_id or not model or not isinstance(messages, list) or not messages:
            continue
        temperature = obj.get("temperature")
        max_tokens = obj.get("max_tokens")
        items.append(
            PromptItem(
                prompt_id=prompt_id,
                model=model,
                messages=messages,
                temperature=float(temperature) if isinstance(temperature, (int, float)) else None,
                max_tokens=int(max_tokens) if isinstance(max_tokens, int) else None,
            )
        )
    return items


def pick_random_prompt(items: list[PromptItem]) -> PromptItem:
    if not items:
        raise ValueError("prompt bank empty")
    return random.choice(items)

File: src/prompt-bank.ts
import { z } from "zod"
import type { PromptItem } from "./types.js"
import { readJsonlFile } from "./util/jsonl.js"

const chatSchema = z.object({
  kind: z.literal("chat").optional(),
  prompt_id: z.string().min(1),
  weight: z.number().int().positive().optional(),
  model: z.string().min(1).optional(),
  messages: z.array(z.unknown()),
  temperature: z.number().optional(),
  max_tokens: z.number().int().positive().optional()
})

const apiKindEnum = z.enum([
  "predexon",
  "search",
  "exa",
  "modal",
  "usstock",
  "stocks",
  "crypto",
  "fx",
  "commodity"
])

const apiSchema = z.object({
  kind: apiKindEnum,
  prompt_id: z.string().min(1),
  weight: z.number().int().positive().optional(),
  method: z.enum(["GET", "POST"]).optional(),
  path: z.string().min(1),
  params: z.record(z.unknown()).optional(),
  body: z.record(z.unknown()).optional()
})

const promptSchema = z.preprocess((v) => {
  if (!v || typeof v !== "object") return v
  const r = v as Record<string, unknown>
  if (typeof r.kind === "string") return v
  return { ...r, kind: "chat" }
}, z.discriminatedUnion("kind", [chatSchema, apiSchema]))

export function loadPromptBank(filePath: string): PromptItem[] {
  return readJsonlFile(filePath, (obj) => promptSchema.parse(obj))
}

export function pickRandomPrompt(bank: PromptItem[]): PromptItem {
  if (!bank.length) throw new Error("empty prompt bank")
  const i = Math.floor(Math.random() * bank.length)
  return bank[i] as PromptItem
}

export function pickWeightedPrompt(bank: PromptItem[]): PromptItem {
  if (!bank.length) throw new Error("empty prompt bank")
  let total = 0
  for (const p of bank) total += typeof (p as any).weight === "number" ? (p as any).weight : 1
  let r = Math.random() * total
  for (const p of bank) {
    r -= typeof (p as any).weight === "number" ? (p as any).weight : 1
    if (r <= 0) return p
  }
  return bank[bank.length - 1] as PromptItem
}
